channels/applicationinsights-channel-js/src/Interfaces.ts

```diff
@@ -29,6 +29,11 @@ export interface ISenderConfig {
      */
     enableSessionStorageBuffer: () => boolean;
 
+    /**
+     * Specify the storage buffer type implementation.
+     */
+    bufferOverride: () => "sessionStorage" | "localStorage" | undefined;
```

Comment:
I was actually thinking that this would be an object with the getStorage / setStorage functions that would be called (if supplied). This would allow anyone to replace with whatever implementation they wanted.

Like some sort of cached / lazy in-memory storage mechanism etc.

Reply:
As there is "consent" issues with storing in LocalStorage that if we supplied we would need to document and follow.

Reply:
> I was actually thinking that this would be an object with the getStorage / setStorage functions that would be called (if supplied)

Yeah... I definitely agree. I started poking at it, but it felt like a much bigger patch, as ideally the exposed interface needs to support async. 

This simple interface seemed like a reasonable first step as a way to support a few more useful first-class storage mechanisms. In a future PR, I'd imagine putting an interface into the type here as well so this becomes:

```
    bufferOverride: () => "sessionStorage" | "localStorage" | ISendBuffer | undefined;
```

> As there is "consent" issues with storing in LocalStorage that if we supplied we would need to document and follow.

Interesting, this hadn't occurred to me.

So, for the specifics here, is your preference this type becomes:
```
    bufferOverride: () => { getItems: StorageGetter: setItems: StorageSetter } | undefined;
```

It'd be trivial to make that happen 🙂 

In future, I'd like to consider exposing the `ISendBuffer` directly, as various other storage solutions (e.g., indexeddb, or a sqlite implementation) probably want the full interface.

Reply:
> `bufferOverride: () => { getItems: StorageGetter: setItems: StorageSetter } | undefined;`

Yes

> In future, I'd like to consider exposing the ISendBuffer directly, as various other storage solutions (e.g., indexeddb, or a sqlite implementation) probably want the full interface.

Assuming that we bring over the same internal batching that we have, it actually won't be necessary as the ISendBuffer will go away entirely as we would be updating Sender to send notifications instead using the existing [`INotificationListener`](https://github.com/microsoft/ApplicationInsights-JS/blob/master/shared/AppInsightsCore/src/JavaScriptSDK.Interfaces/INotificationListener.ts) with the eventsSent, eventsDiscarded, eventsSentRequest.

Part of this is already exposed in the [@microsoft/1ds-post-js](https://www.npmjs.com/package/@microsoft/1ds-post-js?activeTab=code) package, deep in the HttpManager.ts

Reply:
New usage looks something like:
```
const appInsights = new ApplicationInsights({
  config: {
    enableSessionStorageBuffer: true,
    bufferOverride: { 
      getItem: (logger, key) => localStorage.getItem(key),
      setItem: (logger, key, value) => localStorage.setItem(key, value),
    }
  }
});
```

Reply:
Yep, looks good.